proxy/http/HttpTransact.cc

```diff
@@ -4485,7 +4487,21 @@ HttpTransact::handle_cache_operation_on_forward_server_response(State *s)
       }
 
     } else if (s->cache_info.action == CACHE_DO_DELETE) {
-      // do nothing
+      if (!cacheable && is_response_error_code(server_response_code) &&
+          (s->hdr_info.server_request.method_get_wksidx() == HTTP_WKSIDX_DELETE ||
```

Comment:
I think this should omit safe methods, rather than include known unsafe ones. (GET, HEAD, OPTIONS, TRACE; rfc7231) 

HTTP doesn't expressly forbid unknown methods, and methods can be added in a later spec, which would qualify as "a method whose safety is unknown."

Reply:
`is_response_error_code` as-written will also return false for codes >= 600, and true for 1xx codes. No HTTP spec defines a 600+ code, but RFC7234 says "Here, a "non-error response" is one with a 2xx (Successful) or 3xx   (Redirection) status code." 

In this particular case, I think we should follow the spec rather than whatever the client, origin, ATS operator, or ATS programmer configured for general error codes, and check precisely `>= 200 && <= 300`

Reply:
Thank you for the review! These are good observations worth discussing.

Note that the logic this patch adds isn't applying deletion, it's undoing the deletion action in particular circumstances. So the way to apply a careful stance here is to be particular, not broad. Thus I think the concerns you raise are actually satisfied by this patch:

* The methods under consideration are those we know are unsafe, explicitly. If we don't know their safety, then the conditional block is not entered and the resource will be deleted. So for unknown methods, deletion will be preserved. (Or, I should say, if it isn't it's not the fault of this patch!)
* The response codes under consideration are those that are known to be error ones (400-599, inclusive). If they are anything else, then the deletion action will be kept. Thus for 2xx, 1xx, or any theoretical 6xx codes, deletion will be kept.

Related to this: I'm trying to keep the changed behavior narrow. An earlier revision of this patch during my development undid deletion too broadly and broke the splice plugin's assumptions about deletion. Changing caching behavior is a bit risky, so I'm trying to make the changed behavior judiciously narrow while still correct and addressing the linked issue.

Reply:
Talking with Rob offline, we think it would be good to change out the `is_response_error_code` function for `is_response_success_code` and negate that instead. I'll do that and update the PR.

Thanks Rob!

Reply:
Putting a little more of our conversation here for posterity:
This code doesn't introduce spec violations, but it might not let us cache in all scenarios when we could.

Both the method and the codes _could_ be a little broader (including unknown methods, 1xx, 6xx+), but it's not violating anything as-is. And being conservative with behavior changes is probably a good thing.

But overall, I do think this is an improvement and not introducing spec violations. 

I was also asked to verify the `is_response_*_code` changes aren't changing parent or strategy behavior. That part LGTM, behavior appears to be unchanged.

Reply:
Also, not part of this PR, but I believe `does_method_require_cache_copy_deletion` may actually be violating RFC7234 by not returning true for PATCH or unknown methods. I'll try to follow up on that.

Reply:
Thanks again Rob! I appreciate your thoughtful review.